File: src/admin/services/StrapiApiService.ts
import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';

/**
 * 请求参数接口
 */
interface RequestParams {
  [key: string]: any;
}

/**
 * 入口信息接口
 */
interface EntryResponse {
  documentId: string;
  [key: string]: any;
}

/**
 * 语言配置接口
 */
interface Locale {
  code: string;
  name: string;
  isDefault: boolean;
}

/**
 * 获取API Token
 * @returns API Token字符串
 */
export const getApiToken = (): string => {
  const token = process.env.STRAPI_API_TOKEN;
  if (!token) {
    throw new Error('STRAPI_API_TOKEN 环境变量未设置');
  }
  return token;
};

/**
 * Strapi API 服务类 - 处理与 Strapi 后端的所有 API 交互
 */
class StrapiApiService {
  private client: AxiosInstance;
  private token: string;
  private baseUrl: string;

  /**
   * 构造函数 - 初始化 API 客户端
   * @param baseUrl Strapi实例的基础URL，如 "http://localhost:1337"
   * @param token API 令牌，如果不提供，将使用 getApiToken 获取
   */
  constructor(baseUrl: string, token?: string) {
    this.token = token || getApiToken();
    this.baseUrl = baseUrl.replace(/\/$/, ''); // 移除尾部斜杠
    
    // 创建 axios 实例
    this.client = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'Authorization': `Bearer ${this.token}`,
        'Content-Type': 'application/json',
      },
    });
    
    // 添加响应拦截器进行错误处理
    this.client.interceptors.response.use(
      response => response,
      error => {
        console.error('API 请求错误:', error.response?.data || error.message);
        return Promise.reject(error);
      }
    );
  }

  /**
   * 获取 API Token
   * @returns 当前使用的 API Token
   */
  getToken(): string {
    return this.token;
  }

  /**
   * 根据 entry_id 和 locale 获取条目（同时包含草稿和已发布）
   * @param uid 内容类型的路由名（复数），如 "articles"
   * @param entryId 条目的ID
   * @param locale 语言代码，如 "zh-Hant"，默认为 null
   * @param populate 关联字段填充参数，如 "*"，默认为 "*"
   * @returns 条目详细信息或 null
   */
  async getEntry(
    uid: string, 
    entryId: string, 
    locale: string | null = null, 
    populate: string | null = "*"
  ): Promise<any | null> {
    try {
      const url = `/api/${uid}/${entryId}`;
      
      // 构建请求参数
      const params: RequestParams = {
        status: 'draft' // 同时获取草稿和已发布版本
      };
      
      if (locale) {
        params.locale = locale;
      }
      
      if (populate) {
        params.populate = populate;
      }
      
      const response = await this.client.get(url, { params });
      return response.data.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      console.error('获取条目失败:', error);
      throw error;
    }
  }

  /**
   * 发布或保存条目草稿
   * @param uid 内容类型的路由名（复数），如 "articles"
   * @param entryId 条目的ID
   * @param locale 语言代码，如 "zh-Hant"
   * @param data 要更新的数据
   * @param draft 是否为草稿，默认为 true
   * @returns 发布/保存结果
   */
  async publishEntry(
    uid: string, 
    entryId: string, 
    locale: string,
    data: any,
    draft: boolean = true
  ): Promise<any> {
    try {
      const url = `/api/${uid}/${entryId}`;
      
      // 构建请求参数和数据
      const params: RequestParams = {
        locale: locale,
        status: draft ? 'draft' : 'publish'
      };
      
      const payload = {
        data: data
      };
      
      const response = await this.client.put(url, payload, { params });
      return response.data.data?.title;
    } catch (error) {
      console.error('发布/保存条目失败:', error);
      throw error;
    }
  }

  /**
   * 根据 slug 和 locale 获取条目（同时包含草稿和已发布）
   * @param uid 内容类型的路由名（复数），如 "articles"
   * @param slug 条目的 slug
   * @param locale 语言代码，如 "zh-Hant"，默认为 null
   * @param populate 关联字段填充参数，如 "*"，默认为 null
   * @returns 条目详细信息或 null
   */
  async getEntryFromSlug(
    uid: string,
    slug: string,
    locale: string | null = null,
    populate: string | null = null
  ): Promise<any | null> {
    try {
      const url = `/api/${uid}`;
      
      // 构建请求参数
      const params: RequestParams = {
        status: 'draft', // 同时获取草稿和已发布版本
        'filters[slug][$eq]': slug
      };
      
      if (locale) {
        params.locale = locale;
      }
      
      if (populate) {
        params.populate = populate;
      }
      
      const response = await this.client.get(url, { params });
      const items = response.data.data || [];
      return items.length > 0 ? items[0] : null;
    } catch (error) {
      console.error('根据slug获取条目失败:', error);
      throw error;
    }
  }

  /**
   * 获取所有支持的语言配置
   * @returns 语言代码列表
   */
  async getAllLocales(): Promise<string[]> {
    try {
      const url = `/api/i18n/locales`;
      const response = await this.client.get(url);
      return response.data.map((locale: any) => locale.code);
    } catch (error) {
      console.error('获取语言配置失败:', error);
      throw error;
    }
  }
}

export default StrapiApiService; 

File: src/admin/services/StrapiApiService.py
# coding:utf-8
# author:Wayne
# date:2023-xx-xx
import requests
from typing import Optional, Dict, Any


class StrapiApiService(object):
    """
    A simple Python client for Strapi v5 REST API.
    Supports content retrieval, creation, update, and publishing.
    """

    def __init__(self, base_url: str, jwt: str):
        """
        Initialize the service with Strapi base URL and JWT token.

        Args:
            base_url (str): The base URL of the Strapi instance, e.g. "http://localhost:1337".
            jwt (str): A valid JWT for Strapi API (from /auth/local).
        """
        self.base_url = base_url.rstrip('/')
        self.headers = {
            "Authorization": f"Bearer {jwt}",
            "Content-Type": "application/json"
        }

    def get_entry(
            self,
            uid: str,
            entry_id: str,
            locale: Optional[str] = None,
            populate: Optional[str] = "*"
    ) -> Optional[Dict[str, Any]]:
        """
        根据 entry_id 和 locale 获取条目（同时包含草稿和已发布）。

        使用详情接口并添加 publicationState=preview 参数即可同时获取草稿和已发布版本。

        Args:
            uid (str): 内容类型的路由名（复数），如 "articles"。
            entry_id (int): 条目数字 ID。
            locale (Optional[str]): 语言代码，如 "zh-Hant"。
            populate (Optional[str]): 关联字段填充参数，如 "*"。

        Returns:
            Optional[Dict[str, Any]]: 返回条目 data 或 None。
        """
        url = f"{self.base_url}/api/{uid}/{entry_id}?status=draft"
        params: Dict[str, Any] = {}
        if locale:
            params['locale'] = locale
        if populate:
            params['populate'] = populate

        resp = requests.get(url, headers=self.headers, params=params)
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        return resp.json().get('data')

    def publish_entry(
            self,
            uid: str,
            entry_id: str,
            locale: str,
            data: Dict[str, Any],
            draft: bool = True,
    ) -> Dict[str, Any]:

        """
        Publish a draft entry (requires draftAndPublish enabled).

        Args:
            uid (str): The content-type UID.
            entry_id (int): The numeric ID of the entry.
            locale (Optional[str]): I18N locale code to target.
            data:data for update
            draft: publish or just draft

        Returns:
            Dict[str, Any]: The JSON response from Strapi.
        """
        url = f"{self.base_url}/api/{uid}/{entry_id}"
        params = dict()
        params['locale'] = locale

        if draft:
            params['status'] = 'draft'
        else:
            params['status'] = 'publish'
        payload = {
            "data": data
        }

        response = requests.put(url, headers=self.headers, params=params, json=payload)
        response.raise_for_status()
        return response.json().get('data').get('title')

    def get_entry_from_slug(
            self,
            uid: str,
            slug: str,
            locale: Optional[str] = None,
            populate: Optional[str] = None
    ) -> Optional[Dict[str, Any]]:
        """
        根据 slug 和 locale 获取条目（同时包含草稿和已发布）。

        Args:
            uid (str): 内容类型的路由名（复数），如 "articles"。
            slug (str): 条目的 slug。
            locale (Optional[str]): 语言代码，如 "zh-Hant"。
            populate (Optional[str]): 关联字段填充参数，如 "*"。

        Returns:
            Optional[Dict[str, Any]]: 返回条目 data 或 None。
        """
        url = f"{self.base_url}/api/{uid}?status=draft"
        # url = f"{self.base_url}/api/{uid}?filters[slug][$eq]={slug}&populate=*"
        params: Dict[str, Any] = {
            # status=draft 可同时获取草稿和已发布版本
            'status': 'draft',
            # 过滤 slug
            'filters[slug][$eq]': slug
        }
        if locale:
            params['locale'] = locale
        if populate:
            params['populate'] = populate

        resp = requests.get(url, headers=self.headers, params=params)
        resp.raise_for_status()
        items = resp.json().get('data', [])
        return items[0] if items else None

    def get_all_locales(self):
        """
        Fetch all supported locales configured in Strapi.

        Returns:
            List[Dict[str, Any]]: A list of locale objects (code, name, isDefault, etc.).
        """
        url = f"{self.base_url}/api/i18n/locales"
        response = requests.get(url, headers=self.headers)
        response.raise_for_status()

        return [i.get('code') for i in response.json()]


if __name__ == '__main__':
    JWT = '3dbb23dd3d19da0d24cc29a46c5ab7a73d242a3e2054f3ca6b932f9bf50c2b5bb32cde3ab53412c5f8dfd2dceb5aa1b4e03efaa27e62c27d0312443bb9875c6b5b72e3f47237148dcea1bd0b7dbc74b67064f54ef56897d5ec0b6376a8e44339b398e0930161d7292bcb16b279a4ed9721cc44988344474c0f1a12e98389cd88'
    base_url = 'http://127.0.0.1:1337'
    service = StrapiApiService(base_url=base_url, jwt=JWT)

    # category_id = 'd06baauqntszupo9h0f7f44w'
    # article_id = 'i9fi5aekugf63imc8tbbcl4s'
    # # res = service.get_entry(uid='articles', entry_id=article_id, locale='en')
    # # print(res)
    #
    # res = service.get_entry_from_slug(slug='bitcoin', locale='en', uid='articles')
    # print(res)

    all_locales = service.get_all_locales()
    print(all_locales)

# Example usage:
# service = StrapiApiService("http://localhost:1337", jwt="YOUR_JWT_HERE")
# article = service.get_entry("api::article.article", 1, locale="en", populate="*")
# print(article)
# new_entry = service.create_entry("api::article.article", {"title": "Hello"}, locale="en", draft=True)
# updated = service.update_entry("api::article.article", 2, {"title": "Updated"})
# published = service.publish_entry("api::article.article", 2)
